### ISSUE TITLE
[core] deadlock between `poll_all_devices` and `queue_submit`

### ISSUE THREAD

**Issue by sagudev** [gfx-rs/wgpu#5695]
**Description**
Deadlock between `poll_all_devices` (while `deferred_resource_destruction` on bind group destroy) and `queue_submit` (while `device.maintain`):
```
thread backtrace
  thread #111, name = 'WGPU'
    frame #0: 0x00007ffff6f2725d libc.so.6`syscall at syscall.S:38
    frame #1: 0x000055555bd38887 servo`parking_lot::raw_mutex::RawMutex::lock_slow at linux.rs:112:13
    frame #2: 0x000055555bd3886a servo`parking_lot::raw_mutex::RawMutex::lock_slow [inlined] <parking_lot_core::thread_parker::imp::ThreadParker as parking_lot_core::thread_parker::ThreadParkerT>::park at linux.rs:66:13
    frame #3: 0x000055555bd38864 servo`parking_lot::raw_mutex::RawMutex::lock_slow at parking_lot.rs:635:36
    frame #4: 0x000055555bd38809 servo`parking_lot::raw_mutex::RawMutex::lock_slow at parking_lot.rs:207:5
    frame #5: 0x000055555bd38809 servo`parking_lot::raw_mutex::RawMutex::lock_slow at parking_lot.rs:600:5
    frame #6: 0x000055555bd38809 servo`parking_lot::raw_mutex::RawMutex::lock_slow(self=0x00007ffe9db04cf0, timeout=Instant>{...}) at raw_mutex.rs:262:17
    frame #7: 0x000055555b0142d4 servo`core::ptr::drop_in_place<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>> [inlined] <parking_lot::raw_mutex::RawMutex as lock_api::mutex::RawMutex>::lock(self=0x00007ffe9db04cf0) at raw_mutex.rs:72:13
    frame #8: 0x000055555b0142b5 servo`core::ptr::drop_in_place<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>> at mutex.rs:223:9
    frame #9: 0x000055555b0142b5 servo`core::ptr::drop_in_place<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>> [inlined] wgpu_core::lock::vanilla::Mutex<T>::lock at vanilla.rs:29:27
    frame #10: 0x000055555b0142ac servo`core::ptr::drop_in_place<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>> [inlined] <wgpu_core::resource::DestroyedBuffer<A> as core::ops::drop::Drop>::drop(self=0x00007ffea941b150) at resource.rs:646:57
    frame #11: 0x000055555b0142a4 servo`core::ptr::drop_in_place<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>((null)=0x00007ffea941b150) at mod.rs:497:1
    frame #12: 0x000055555b01855d servo`alloc::sync::Arc<T,A>::drop_slow(self=<unavailable>) at sync.rs:1751:18
    frame #13: 0x000055555b102c80 servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] <alloc::sync::Arc<T,A> as core::ops::drop::Drop>::drop(self=<unavailable>) at sync.rs:2407:13
    frame #14: 0x000055555b102c7b servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::drop_in_place<alloc::sync::Arc<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>>((null)=<unavailable>) at mod.rs:497:1
    frame #15: 0x000055555b102c7b servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::drop_in_place<(wgpu_core::track::TrackerIndex,alloc::sync::Arc<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>)>((null)=<unavailable>) at mod.rs:497:1
    frame #16: 0x000055555b102c6d servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::mut_ptr::<impl *mut T>::drop_in_place at mut_ptr.rs:1461:18
    frame #17: 0x000055555b102c6d servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] hashbrown::raw::Bucket<T>::drop(self=<unavailable>) at mod.rs:581:23
    frame #18: 0x000055555b102c6d servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> at mod.rs:1038:17
    frame #19: 0x000055555b102bee servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] <hashbrown::raw::RawTable<T,A> as core::ops::drop::Drop>::drop(self=0x00007fff447f7480) at mod.rs:2699:17
    frame #20: 0x000055555b102bde servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::drop_in_place<hashbrown::raw::RawTable<(wgpu_core::track::TrackerIndex,alloc::sync::Arc<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>)>>((null)=0x00007fff447f7480) at mod.rs:497:1
    frame #21: 0x000055555b102bde servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::drop_in_place<hashbrown::map::HashMap<wgpu_core::track::TrackerIndex,alloc::sync::Arc<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>,core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>((null)=0x00007fff447f7480) at mod.rs:497:1
    frame #22: 0x000055555b102bde servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>> [inlined] core::ptr::drop_in_place<std::collections::hash::map::HashMap<wgpu_core::track::TrackerIndex,alloc::sync::Arc<wgpu_core::resource::DestroyedBuffer<wgpu_hal::vulkan::Api>>,core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>((null)=size=1, capacity=4) at mod.rs:497:1
    frame #23: 0x000055555b102bde servo`core::ptr::drop_in_place<wgpu_core::device::life::ResourceMaps<wgpu_hal::vulkan::Api>>((null)=0x00007fff447f7300) at mod.rs:497:1
    frame #24: 0x000055555b111172 servo`wgpu_core::device::life::LifetimeTracker<A>::triage_submissions(self=0x00007ffe9db04888, last_done=<unavailable>, command_allocator=0x00007ffe9db04538) at life.rs:413:9
    frame #25: 0x000055555b0d5d08 servo`wgpu_core::device::resource::Device<A>::maintain(self=0x00007ffe9db02010, fence_guard=wgpu_core::lock::vanilla::RwLockReadGuard<core::option::Option<wgpu_hal::vulkan::Fence>> @ r13, maintain=<unavailable>, snatch_guard=<unavailable>) at resource.rs:436:13
    frame #26: 0x000055555b075ca9 servo`wgpu_core::device::queue::<impl wgpu_core::global::Global>::queue_submit(self=<unavailable>, queue_id=<unavailable>, command_buffer_ids=<unavailable>) at queue.rs:1494:23
```

```
thread backtrace
  thread #112, name = 'WGPU poller'
    frame #0: 0x00007ffff6f2725d libc.so.6`syscall at syscall.S:38
    frame #1: 0x000055555bd36cd2 servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers at linux.rs:112:13
    frame #2: 0x000055555bd36cb5 servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers [inlined] <parking_lot_core::thread_parker::imp::ThreadParker as parking_lot_core::thread_parker::ThreadParkerT>::park at linux.rs:66:13
    frame #3: 0x000055555bd36c96 servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers at parking_lot.rs:635:36
    frame #4: 0x000055555bd36b1b servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers at parking_lot.rs:207:5
    frame #5: 0x000055555bd36a85 servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers at parking_lot.rs:600:5
    frame #6: 0x000055555bd36a85 servo`parking_lot::raw_rwlock::RawRwLock::wait_for_readers(self=0x00007ffe9db045a8, timeout=Instant>{...}, prev_value=0) at raw_rwlock.rs:1017:17
    frame #7: 0x000055555bd3409e servo`parking_lot::raw_rwlock::RawRwLock::lock_exclusive_slow(self=0x00007ffe9db045a8, timeout=Instant>{...}) at raw_rwlock.rs:647:9
    frame #8: 0x000055555b176db3 servo`wgpu_core::device::resource::Device<A>::deferred_resource_destruction [inlined] <parking_lot::raw_rwlock::RawRwLock as lock_api::rwlock::RawRwLock>::lock_exclusive(self=0x00007ffe9db045a8) at raw_rwlock.rs:73:26
    frame #9: 0x000055555b176d97 servo`wgpu_core::device::resource::Device<A>::deferred_resource_destruction at rwlock.rs:500:9
    frame #10: 0x000055555b176d97 servo`wgpu_core::device::resource::Device<A>::deferred_resource_destruction [inlined] wgpu_core::lock::vanilla::RwLock<T>::write at vanilla.rs:85:33
    frame #11: 0x000055555b176d97 servo`wgpu_core::device::resource::Device<A>::deferred_resource_destruction [inlined] wgpu_core::snatch::SnatchLock::write(self=0x00007ffe9db045a8) at snatch.rs:154:40
    frame #12: 0x000055555b176d7f servo`wgpu_core::device::resource::Device<A>::deferred_resource_destruction(self=0x00007ffe9db02010) at resource.rs:359:70
    frame #13: 0x000055555b1640d0 servo`wgpu_core::device::global::<impl wgpu_core::global::Global>::poll_all_devices at global.rs:2152:9
    frame #14: 0x000055555b163fe9 servo`wgpu_core::device::global::<impl wgpu_core::global::Global>::poll_all_devices at global.rs:2188:21
    frame #15: 0x000055555b163e9f servo`wgpu_core::device::global::<impl wgpu_core::global::Global>::poll_all_devices(self=<unavailable>, force_wait=<unavailable>) at global.rs:2213:17
    frame #16: 0x000055555b051c66 servo`std::sys_common::backtrace::__rust_begin_short_backtrace [inlined]
```

**Repro steps**
Similar to https://github.com/gfx-rs/wgpu/issues/5687, but this one is way harder to reproduce (happens more rarely). Using servo: https://github.com/servo/servo/pull/32266/commits/258406ddbdef5cf526a5a7f83e0fb7871470ec73 and running `https://gpuweb.github.io/cts/standalone/?runnow=1&q=webgpu:api,operation,compute,basic:*`.

**Expected vs observed behavior**
Expected no deadlock.

**Platform**
wgpu-core on trunk (d0a5e48aa7e84683114c3870051cc414ae92ac03)

**Comment by sagudev** [comment]
One thread tries to acquire lock on https://github.com/gfx-rs/wgpu/blob/fa48562229ae9effe10696e74249304f1fb9a3f0/wgpu-core/src/resource.rs#L646 while other already acquired it: https://github.com/gfx-rs/wgpu/blob/fa48562229ae9effe10696e74249304f1fb9a3f0/wgpu-core/src/device/resource.rs#L335
this second thread tries to acquire: https://github.com/gfx-rs/wgpu/blob/fa48562229ae9effe10696e74249304f1fb9a3f0/wgpu-core/src/device/resource.rs#L359
that is already acquired by https://github.com/gfx-rs/wgpu/blob/fa48562229ae9effe10696e74249304f1fb9a3f0/wgpu-core/src/device/queue.rs#L1162 and later passed on to https://github.com/gfx-rs/wgpu/blob/fa48562229ae9effe10696e74249304f1fb9a3f0/wgpu-core/src/device/queue.rs#L1494

**Comment by ArthurBrussee** [comment]
I _think_ I am hitting this too:

```
[2024-12-26T22:09:10Z ERROR wgpu_hal::vulkan::instance]         objects: (type: COMMAND_BUFFER, hndl: 0x1f860b4d0e0, name: (wgpu internal) Pre Pass), (type: COMMAND_POOL, hndl: 0x831db900000044b7, name: ?)
[2024-12-26T22:09:10Z ERROR wgpu_hal::vulkan::instance] VALIDATION [VUID-vkResetCommandPool-commandPool-00040 (0xb53e2331)]
        Validation Error: [ VUID-vkResetCommandPool-commandPool-00040 ] Object 0: handle = 0x1f860b4b7d0, name = CubeCL Command Encoder, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0x831db900000044b7, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0xb53e2331 | vkResetCommandPool():  (VkCommandBuffer 0x1f860b4b7d0[CubeCL Command Encoder]) is in use.
    The Vulkan spec states: All VkCommandBuffer objects allocated from commandPool must not be in the pending state (https://vulkan.lunarg.com/doc/view/1.3.296.0/windows/1.3-extensions/vkspec.html#VUID-vkResetCommandPool-commandPool-00040)
[2024-12-26T22:09:10Z ERROR wgpu_hal::vulkan::instance]         objects: (type: COMMAND_BUFFER, hndl: 0x1f860b4b7d0, name: CubeCL Command Encoder), (type: COMMAND_POOL, hndl: 0x831db900000044b7, name: ?)
[2024-12-26T22:09:10Z ERROR wgpu_hal::vulkan::instance] VALIDATION [VUID-VkSubmitInfo-pSignalSemaphores-03244 (0x7fd092df)]
        Validation Error: [ VUID-VkSubmitInfo-pSignalSemaphores-03244 ] Object 0: handle = 0xf56c9b0000000004, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x1f808884830, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x7fd092df | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] value (28863) exceeds limit regarding current semaphore VkSemaphore 0xf56c9b0000000004[] value (18446744073709551615).
    The Vulkan spec states: For each element of pSignalSemaphores created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_TIMELINE the corresponding element of VkTimelineSemaphoreSubmitInfo::pSignalSemaphoreValues must have a value which does not differ from the current value of the semaphore or the value of any outstanding semaphore wait or signal operation on that semaphore by more than maxTimelineSemaphoreValueDifference (https://vulkan.lunarg.com/doc/view/1.3.296.0/windows/1.3-extensions/vkspec.html#VUID-VkSubmitInfo-pSignalSemaphores-03244)
[2024-12-26T22:09:10Z ERROR wgpu_hal::vulkan::instance]         objects: (type: SEMAPHORE, hndl: 0xf56c9b0000000004, name: ?), (type: QUEUE, hndl: 0x1f808884830, name: ?)
thread 'tokio-runtime-worker' panicked at C:\Users\A-Bru\Documents\wgpu\wgpu-core\src\resource.rs:704:59:
thread 'tokio-runtime-worker' attempted to acquire a snatch lock recursively.
- Currently trying to acquire a write lock at C:\Users\A-Bru\Documents\wgpu\wgpu-core\src\resource.rs:704:59
disabled backtrace
- Previously acquired a read lock at C:\Users\A-Bru\Documents\wgpu\wgpu-core\src\device\queue.rs:1090:60
disabled backtrace
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2024-12-26T22:12:28Z ERROR wgpu::backend::wgpu_core] Handling wgpu errors as fatal by default
thread 'main' panicked at C:\Users\A-Bru\Documents\wgpu\wgpu\src\backend\wgpu_core.rs:1529:26:
wgpu error: Validation Error

Caused by:
  In Device::create_command_encoder, label = 'encoder'
    Device with 'egui+burn' label is invalid.
```

This seems to happen while some bind group is released, while waiting for some async mapping result to come in.

Tested at both 23, 23.1, and at head.

**Comment by andyleiserson** [comment]
Here is Claude's analysis concluding that this deadlock no longer exists after the changes in #9479:

## 1. Is `poll_and_return_closures` the equivalent of `poll_all_devices` in the issue?

For the deferred-destruction logic, yes. In the old code (commit d0a5e48aa7e8), `poll_all_devices` itself called `device.maintain(...)` and then `device.deferred_resource_destruction()` for each device. That work has since moved into `Device::poll_and_return_closures` (`wgpu-core/src/device/resource.rs:796–811`). Both `Global::poll_all_devices_of_api` and `Global::device_poll` now go through it (`global.rs:1871, 1912`).

## 2. Is the deadlock no longer possible?

Yes — and **two** independent fixes break the cycle, not one:

**Your fix (Side A)**: `Device::maintain` now explicitly drops the `fence` and `snatch_guard` at `resource.rs:938–942` before returning, so when `poll_and_return_closures` calls `deferred_resource_destruction()` at line 808, neither lock is held.

**A separate fix (Side B)**: `deferred_resource_destruction` itself no longer holds the `deferred_destroy` mutex across the `snatchable_lock.write()` call. The old code at d0a5e48aa7e8 was:

```rust
while let Some(item) = self.deferred_destroy.lock().pop() {
    ...
    let Some(raw_view) = view.raw.snatch(self.snatchable_lock.write()) else { ... };
```

In `while let`, temporaries in the condition live to the end of the loop body, so the `deferred_destroy` MutexGuard was held while `snatchable_lock.write()` was acquired. That's the "Thread B" side of the original cycle. The current code (`resource.rs:731–732`) is:

```rust
let deferred_destroy = mem::take(&mut *self.deferred_destroy.lock());
for item in deferred_destroy {
```

The temporary guard is dropped at the `;`. The loop iterates over an owned `Vec` and acquires `snatchable_lock.write()` with `deferred_destroy` no longer held.

Either fix alone is sufficient. The two together make this defense-in-depth.

## 3. Other DEVICE_DEFERRED_DESTROY codepaths

There are only three acquisition sites for the lock:

- `DestroyedBuffer::drop` — `resource.rs:1140`
- `DestroyedTexture::drop` — `resource.rs:1606`
- `Device::deferred_resource_destruction` — `resource.rs:731`

The first two only push into the `Vec` and release; they never try to acquire any other wgpu-core lock while holding `deferred_destroy`. The third releases `deferred_destroy` before acquiring anything else (as shown above).

That means **no thread ever waits on another wgpu-core lock while holding `DEVICE_DEFERRED_DESTROY`**, so this lock cannot be on the "blocked-on" side of any deadlock cycle. The class of bug from #5695 is structurally impossible regardless of what code paths are exercised.

The remaining concern is lock-*ordering* (rank-validator) violations rather than deadlocks: any path that drops a `DestroyedBuffer`/`DestroyedTexture` while holding a lock whose follower set doesn't transitively include `DEVICE_DEFERRED_DESTROY` would fail validation. Per the rank file, the predecessors that reach `DEVICE_DEFERRED_DESTROY` are `DEVICE_TRACKERS`, `QUEUE_LIFE_TRACKER`, `DEVICE_COMMAND_INDICES` (direct), plus anything that reaches one of those (e.g., `DEVICE_SNATCHABLE_LOCK`, `DEVICE_FENCE`, `QUEUE_PENDING_WRITES`). Locks like `COMMAND_BUFFER_DATA`, `BUFFER_MAP_STATE`, `BLAS_COMPACTION_STATE`, all the `REGISTRY_*` ranks, and the leaf locks do NOT reach it. So unexercised paths to watch for would be ones that drop a `TempResource::DestroyedBuffer`/`DestroyedTexture` (or `Arc<DestroyedBuffer>` if any survive) while holding one of those — most plausibly error/early-return paths in `queue_submit`, `Buffer::destroy`/`Texture::destroy` failure scenarios, and command-encoder error paths that drop `temp_resources`. I checked `Buffer::destroy` (`resource.rs:1046–1072`): in the no-`last_submit_index` branch, `life_lock` is declared after the rebound `temp`, so `life_lock` drops before `temp`, and `DestroyedBuffer::drop` runs with no wgpu-core locks held — that path is safe.
